src/basic/user-util.c

```diff
@@ -553,14 +550,7 @@ int getgroups_alloc(gid_t** gids) {
                 ngroups = getgroups(0, NULL);
                 if (ngroups < 0)
                         return -errno;
-                if (ngroups == 0)
-                        return false;
```

Comment:
Still, this deviates from our coding style. We prefer setting ret param to NULL if effectively caller would get nothing, rather than an empty buffer.

Reply:
I won't do it in this PR for two reasons:

-  This is a bug fix, not a style fix.  I think bug fixes should not be merged with style fixes.  Since the normal run in which we return success in the first iteration does set the return output to a pointer to unused memory (as if malloc(1)), I'll do the same exact thing here.

-  Do you really use NULL as a 0-size buffer?  I see you have the `?:1` good thing in malloc(3) calls to make sure that you don't have NULLs messing around.  Are you sure you want a NULL here?

```c
alx@devuan:~/src/systemd/systemd/agetgroups$ find -type f | grep '\.[ch]$' | xargs grepc new
./src/basic/alloc-util.h:#define new(t, n) ((t*) malloc_multiply(n, sizeof(t)))
alx@devuan:~/src/systemd/systemd/agetgroups$ sed -n 115,120p src/basic/alloc-util.h
_malloc_ _alloc_(1, 2) static inline void *malloc_multiply(size_t need, size_t size) {
        if (size_multiply_overflow(size, need))
                return NULL;

        return malloc(size * need ?: 1);
}
```

Reply:
yes, we generally try to zero-sized arrays and NULL the same with very few exceptions.

malloc_multiply() bumps the size to 1 at least simply so that we always can reasonably distinguish OOM from a broken malloc implementation that is stupid enough to return NULL for zero sized allocations.

libsystemd is sometimes loaded into weird processes (by means of NSS modules and such), which interpose malloc(), and we kinda want to avoid an confusion

Reply:
Still, a zero-size buffer is better in this case, because it removes a branch.  I don't see how adding a special case for zero setting the pointer to NULL is beneficial at all.

But yeah, this would be less egreggious than the brain-damaged decision of allowing malloc(0) to return NULL.